`NSCoding`と`NSKeyedUnarchiver`を使ってファイルに保存していたのを、`Codable`を使った方法に変更

hhyyg opened the pull request:
`NSCoding`と`NSKeyedUnarchiver`を使ってファイルに保存していたのを、`Codable`を使った方法に変更

hhyyg committed:
[using Codable] UIImageWrapperの追加

hhyyg committed:
[using Codable] FileManagerでの保存処理を変更

hhyyg committed:
import Log.swift

mono0926 commented on Miso.TrackingMealApp/TrackingMealAppTests/UIImageWrapperTest.swift:
```diff
@@ -0,0 +1,51 @@
+            // Put the code you want to measure the time of here.
+        }
+    }
+
```
不要な行は消しましょう。

mono0926 commented on Miso.TrackingMealApp/TrackingMealAppTests/UIImageWrapperTest.swift:
```diff
@@ -0,0 +1,51 @@
+
+        //do
+        _ = try! JSONDecoder().decode(UIImageWrapper.self, from: encoded)
+    }
```
encode/decodeして、`#imageLiteral(resourceName: "filledStar")`とequalになるか検証が良さそう。
(そもそもそれでequal扱いになるかちょっと自信無いけど)

mono0926 replied:
> (そもそもそれでequal扱いになるかちょっと自信無いけど)

画像はUIImagePNGRepresentationで可逆変換になってるものの、微妙 🤔

hhyyg replied:
不要になったので、テスト自体削除しました。

mono0926 commented on Miso.TrackingMealApp/TrackingMealApp/AppError.swift:
```diff
@@ -0,0 +1,13 @@
+
+import Foundation
+
+struct AppError: Error {
```
enumで、decodeFailedとか種別用意した方が良い。

hhyyg replied:
なるほど。整理したらAppError不要になったので削除しました。

mono0926 commented on Miso.TrackingMealApp/TrackingMealApp/Log.swift:
```diff
@@ -0,0 +1,132 @@
+//
+//  Created by mono on 2017/02/11.
+//  Copyright © 2017 mono. All rights reserved.
+//
```
mono発見(　´･?･｀)

hhyyg replied:
Bundle.main.bundleIdentifier に変更して使いました

mono0926 commented on Miso.TrackingMealApp/TrackingMealApp/Log.swift:
```diff
@@ -0,0 +1,132 @@
+        case OSLogType.error:
+            return "‼️(error)"
+        case OSLogType.fault:
+            return "💣(fault)"
```
絵文字、赤系の目立つ色が良かったと後悔中。

mono0926 commented on Miso.TrackingMealApp/TrackingMealApp/MealTableViewController.swift:
```diff
@@ -139,42 +139,39 @@ class MealTableViewController: UITableViewController {
+            print("url: \(Meal.archiveUrl)")
+            logger.debug("Meals successfully saved")
+        } catch {
+            fatalError(error.localizedDescription)
```
なんでlogger使わずにfatalErrorしたか疑問。
このケースでクラッシュさせたいならdo-try使わずにtry!すれば良い。他も同様。

hhyyg replied:
意味はなく。整理しました。

mono0926 commented on Miso.TrackingMealApp/TrackingMealApp/MealTableViewController.swift:
```diff
@@ -139,42 +139,39 @@ class MealTableViewController: UITableViewController {
         }
     }
 
     private func loadMeals() -> [Meal]? {
```
このあたり別クラスに分けたほうが良いね。

hhyyg replied:
分けました。

mono0926 commented on Miso.TrackingMealApp/TrackingMealApp/UIImageWrapper.swift:
```diff
@@ -0,0 +1,42 @@
+public struct UIImageWrapper: Codable {
+    public let image: UIImage
+
+    public enum CodingKeys: String, CodingKey {
```
多分無しで動く。

mono0926 commented on Miso.TrackingMealApp/TrackingMealApp/UIImageWrapper.swift:
```diff
@@ -0,0 +1,42 @@
+import UIKit
+
+public struct UIImageWrapper: Codable {
+    public let image: UIImage
```
`imageData: Data`のプロパティを持つ構造(イニシャライザーあたりでUIImageからDataに変換)にしたら、下の`public init(from decoder: Decoder) throws`とかの記述がみんな不要になるはず。

hhyyg replied:
すごく整理できました。

mono0926 reviewed (changes requested):
特にCodable周りをずっとスッキリ書けそう

hhyyg committed:
ファイル保存処理を切り分け

tryとfatalErrorの整理
ファイル保存処理を切り分け

hhyyg committed:
Meal.photoをUIImageWrapper型からData型に変更

hhyyg committed:
不要なものを削除

hhyyg committed:
AppError使ってないので削除

hhyyg commented:
@mono0926 修正しました

mono0926 commented on Miso.TrackingMealApp/TrackingMealApp/FileStorage.swift:
```diff
@@ -0,0 +1,27 @@
+import Foundation
+
+class FileStorage {
+    static func store<T>(atPath pathString: String, value: T) throws where T: Encodable {
```
where使わずに`store<T: Encodable>`で書けるはずで、これで済む場合はこっちの方がシンプル・見やすいので推奨かも。

hhyyg replied:
なるほど修正しました

hhyyg committed:
Genericsの書き方を変更

mono0926 commented on Miso.TrackingMealApp/TrackingMealApp/FileStorage.swift:
```diff
@@ -0,0 +1,27 @@
+class FileStorage {
+    static func store<T>(atPath pathString: String, value: T) throws where T: Encodable {
+        let data = try JSONEncoder().encode(value)
+        if FileManager.default.fileExists(atPath: pathString) {
```
`let fm = FileManager.default`して`fm`を使う方が書きやすい。

mono0926 replied:
> let fm = FileManager.defaultしてfmを使う方が書きやすい。

2回以上使う場合。

hhyyg replied:
`fm`は慣習。スコープも少ないので、fm に変更しました。

mono0926 commented on Miso.TrackingMealApp/TrackingMealApp/FileStorage.swift:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    static func retrive<T>(atPath pathString: String) throws -> T? where T: Decodable {
+        if let data = FileManager.default.contents(atPath: pathString) {
```
戻り値がOptionalかつthrowsなことに違和感。
どっちかに寄せた方が良い。

hhyyg replied:
- ファイルが存在しない場合は、正常系としてありえるので、nil を返したい
- エラーが発生する場合は落としたい
という意図がありました。
この場合は、エラーは発生した時点で落としてよかったので、FileStorage内で try! に変更しました。
結果、他のところで try が少なくなりスッキリしたと思います。

mono0926 commented on Miso.TrackingMealApp/TrackingMealApp/MealTableViewController.swift:
```diff
@@ -17,7 +17,7 @@ class MealTableViewController: UITableViewController {
 
         navigationItem.leftBarButtonItem = editButtonItem
-        if let savedMeals = loadMeals() {
+        if let savedMeals = try! loadMeals() {
```
> 戻り値がOptionalかつthrowsなことに違和感。

使う側も、こういう風にtry!しつつif letでアンラップという不自然な感じに。

mono0926 commented on Miso.TrackingMealApp/TrackingMealApp/FileStorage.swift:
```diff
@@ -0,0 +1,27 @@
+        FileManager.default.createFile(atPath: pathString, contents: data, attributes: nil)
+    }
+
+    static func retrive<T>(atPath pathString: String) throws -> T? where T: Decodable {
```
> `(atPath pathString: String)`

冗長。
`(at path: String)`が良いはず。

hhyyg replied:
URL型に変更して同時に直しました。

hhyyg committed:
データファイルの保存場所を表す変数の型を、String から URL に変更

hhyyg committed:
try! をなるべく減らす

- fileManger -> fm
- ファイルからオブジェクトへの変換（またその逆は）はその時点でエラーで落とす
- ファイルがない場合は、nil を返す

hhyyg commented:
@mono0926 修正しました

hhyyg committed:
スコープを最低限に

mono0926 commented:
![02_](https://user-images.githubusercontent.com/1255062/33597003-5493a3f2-d9e1-11e7-9825-ff0a3759afb9.png)

hhyyg commented:
![38_](https://user-images.githubusercontent.com/8636660/33597071-931403b0-d9e1-11e7-85b1-9a43efa5752c.jpg)
